File: 1-Changing_Intent/stats.ts
class Statistics {
    private samples: number[] = [];

    add(value: number): void {
        this.samples.push(value);
    }

    getMean(): number | null {
        if (this.samples.length === 0)
            return null;

        let sum = 0;
        for (const sample of this.samples)
            sum += sample;
        return sum / this.samples.length;
    }

    getStandardDeviation(): number | null {
        const mean = this.getMean();
        if (mean === null)
            return null;

        let sum = 0;
        for (const sample of this.samples) {
            sum += (sample - mean) * (sample - mean);
        }
        return Math.sqrt(sum / this.samples.length - 1);
    }

    getMin(): number | null {
        if (this.samples.length === 0)
            return null;

        let min = this.samples[0];
        for (const sample of this.samples) {
            if (sample < min) {
                min = sample;
            }
        }
        return min;
    }
}

File: 1-Changing_Intent/stats.rb
class Statistics
  def initialize
    @samples = []
  end

  def add(value)
    @samples << value
  end

  def get_mean
    return nil if @samples.empty?

    sum = @samples.sum
    sum.to_f / @samples.size
  end

  def get_standard_deviation
    mean = get_mean
    return nil if mean.nil?

    sum = @samples.reduce(0) do |acc, sample|
      acc + (sample - mean) ** 2
    end
    Math.sqrt(sum / @samples.size - 1)
  end

  def get_min
    return nil if @samples.empty?

    @samples.min
  end
end
